dart/lib/src/reference.dart

```diff
@@ -0,0 +1,416 @@
+    }
+    var byteData = ByteData.view(buffer);
+    var byteWidth = byteData.getUint8(len - 1);
+    var packedType = byteData.getUint8(len - 2);
+    var offset = len - byteWidth - 2;
+    return Reference._(ByteData.view(buffer), offset, BitWidthUtil.fromByteWidth(byteWidth), packedType);
+  }
+
+  bool get isNull => _valueType == ValueType.Null;
+  bool get isNum => ValueTypeUtils.isNumber(_valueType) || ValueTypeUtils.isIndirectNumber(_valueType);
+  bool get isDouble => _valueType == ValueType.Float || _valueType == ValueType.IndirectFloat;
+  bool get isInt => isNum && !isDouble;
+  bool get isString => _valueType == ValueType.String || _valueType == ValueType.Key;
+  bool get isBool => _valueType == ValueType.Bool;
+  bool get isBlob => _valueType == ValueType.Blob;
+  bool get isVector => ValueTypeUtils.isAVector(_valueType);
+  bool get isMap => _valueType == ValueType.Map;
+
+  /// Returns [bool] value or null otherwise.
+  bool get boolValue {
+    if(_valueType == ValueType.Bool) {
+      return _readInt(_offset, _parentWidth) != 0;
+    }
+    return null;
+  }
+
+  /// Returns [int] value or null otherwise.
+  int get intValue {
+    if (_valueType == ValueType.Int) {
+      return _readInt(_offset, _parentWidth);
+    }
+    if (_valueType == ValueType.UInt) {
+      return _readUInt(_offset, _parentWidth);
+    }
+    if (_valueType == ValueType.IndirectInt) {
+      return _readInt(_indirect, BitWidthUtil.fromByteWidth(_byteWidth));
+    }
+    if (_valueType == ValueType.IndirectUInt) {
+      return _readUInt(_indirect, BitWidthUtil.fromByteWidth(_byteWidth));
+    }
+    return null;
+  }
+
+  /// Returns [double] value or null otherwise.
+  double get doubleValue {
+    if (_valueType == ValueType.Float) {
+      return _readFloat(_offset, _parentWidth);
+    }
+    if (_valueType == ValueType.IndirectFloat) {
+      return _readFloat(_indirect, BitWidthUtil.fromByteWidth(_byteWidth));
+    }
+    return null;
+  }
+
+  /// Returns [num] value or null otherwise.
+  num get numValue {
+    if (isDouble) {
+      return doubleValue;
+    }
+    return intValue;
+  }
+
+  /// Returns [String] value or null otherwise.
+  String get stringValue {
+    if (_valueType == ValueType.String || _valueType == ValueType.Key) {
+      return utf8.decode(_buffer.buffer.asUint8List(_indirect, length));
+    }
+    return null;
+  }
+
+  /// Returns [Uint8List] value or null otherwise.
+  Uint8List get blobValue {
+    if (_valueType == ValueType.Blob) {
+      return _buffer.buffer.asUint8List(_indirect, length);
+    }
+    return null;
+  }
+
+  /// Can be used with an [int] or a [String] value for key.
+  /// If the underlying value in FlexBuffer is a vector, then use [int] for access.
+  /// If the underlying value in FlexBuffer is a map, then use [String] for access.
+  /// Returns [Reference] value or null. Does not throw out of bounds exception.
+  Reference operator [](Object key) {
```

Comment:
This seems very likely to introduce silent bugs into programs. It currently declares it accepts one of two types, and returns null if you give it any other type. It also returns null if I try to get something that is out of range.

Is it really necessary? If anything, it seems like it would be better to expose two specific methods for this, and tighten up the contract a little bit so they throw if you do things like go out of range.

Reply:
It is not 100% necessary. I designed it for ease of use, to extract deeply nested values e.g.:

```
final mainCityOfMaximsSecondBestFriend = contacts["Maxim"]["friends"][2]["address"][0]["city"].stringValue;
```

We could also introduce distinguished methods and the expression would looks something like this:
```
final cityOfMaximsBestFriend = contacts.mapValue("Maxim").mapValue("friends").vectorValue(0).mapValue("address").vectorValue(2).mapValue("city").stringValue;
```

In Dart implementation the `Reference` class is completely polymorphic and can represent any kind of value. As matter of fact in C++ implementation there are multiple types of References, which complicates the access pattern. Also in regards to performance and the fact that in Dart we don't have structs, having different references means that we will unnecessary wrap objects in objects which could have an impact on performance.

Reply:
An idea which just popped into my head and which I would like to get feedback on from @aardappel. When having a deep value access like the one I described in previous comment e.g.:
```
final mainCityOfMaximsSecondBestFriend = contacts["Maxim"]["friends"][2]["address"][0]["city"].stringValue;
```

If the path is wrong we get a nothing saying Exception. What if `Reference` would have a `String path` field, which carries the absolute path to this reference. This way in case of a wrong access we can generate much more understandable exception. This however have a downside of Reference having a string field and being less lightweight as it is now.

Reply:
My main concern is that Dart doesn't really have first class support for function (or operator) overloading. The current implementation will make it very easy to introduce bugs that will maybe result in an NPE somewhere further up the call stack from here, and it will be difficult for developers to figure out why, particularly if they just happened to refactor a variable that used to be a string into an object and forgot to get the string from the object before passing it to this method. 

Even just an assertion here that the types are valid would be an improvement, but that's still only a runtime exception on a branch that may not have test coverage. My ideal scenario for this is a method that will be an analysis/compile error to misuse.

Reply:
Fro compile error people will have to use FlatBuffers, FlexBuffers is rather like JSON, the value can be anything and so compiler can't help much. That sad I added very descriptive runtime exceptions now, see:
https://github.com/mzaks/flatbuffers/blob/master/dart/test/flex_reader_test.dart#L242
So developers should be able to understand what caused the issue.

I am resolving this conversation. Please open a new one if you think it should be refined further.